### PR TITLE
Add support for Firefox and Edge

### PR DESCRIPTION
**Pull request by GeeTransit**
I tested Firefox and it works all fine. Not sure about Safari though. The `TyperacerBot` class is removed and its methods are now functions. There could be more prompts for configuration later on.

### CODE REVIEW

File: run.bat [hunk 1]
```diff
@@ -7,4 +7,4 @@ echo installing requirements
 python -m pip install -r requirements.txt
 
 echo starting app
-start python typeracer_bot.py
+python typeracer_bot.py
```
Comment (Togohogo1) [inline run.bat:10]: kinda messy opening in the same window

File: typeracer_bot.py [hunk 2]
```diff
@@ -1,78 +1,81 @@
...
+def send_text(driver, text, delay):
+    """Send `text` to text input box first working driver in `drivers` dict."""
+    # Click the inputbox when it can be clicked
+    elem = WebDriverWait(driver, 15).until(EC.element_to_be_clickable((By.CLASS_NAME, "txtInput")))
+    for i in text:
+        time.sleep(delay)
+        elem.send_keys(i)
+
+# dict[name: lambda -> WebDriver]
+DRIVER_FUNCS = dict(
```
Comment (Togohogo1) [inline typeracer_bot.py:44]: lolz imma move this to the top

File: typeracer_bot.py [hunk 3]
```diff
@@ -1,78 +1,81 @@
...
+    chrome=lambda: Chrome(executable_path="driver/chromedriver"),
+    firefox=lambda: Firefox(executable_path="driver/geckodriver"),
+    safari=lambda: Safari(executable_path="driver/safaridriver"),
+)
+
+def get_driver(preferred=None, drivers=DRIVER_FUNCS):
```
Comment (kwangzera) [inline typeracer_bot.py]: Why not define `DRIVER_FUNCS` in the function and get rid of `drivers=DRIVER_FUNCS`

File: typeracer_bot.py [hunk 4]
```diff
@@ -1,78 +1,81 @@
 import time
```
Comment (kwangzera) [inline typeracer_bot.py:1]: lmao i will reformat ur code my way

File: typeracer_bot.py [hunk 5]
```diff
@@ -1,78 +1,81 @@
...
+
+def get_driver(preferred=None, drivers=DRIVER_FUNCS):
+    """Return first working driver in `drivers` dict."""
+    if preferred is not None:
+        with suppress(WebDriverException):
+            return drivers[preferred]()
+    for name, func in drivers.items():
+        with suppress(WebDriverException):
+            return func()
+    raise LookupError("Driver not found: {preferred=!r}")
```
Comment (kwangzera) [inline typeracer_bot.py]: what does `{preferred=!r}` stand for

File: typeracer_bot.py [hunk 6]
```diff
@@ -1,78 +1,81 @@
...
+    # Open the link
+    driver.get(link)
 
     # Able to run for multiple races
     while True:
         try:
             delay = float(input("Set your delay and press enter when the race starts: "))
         except ValueError:
-            print("- default delay time set to 0.1s")
+            print(f"- using previous delay time of {delay}s")
```
Comment (kwangzera) [inline typeracer_bot.py]: I originally have it default to 0.1s but I forgot to push the version where it had `delay = 0.1` after `print("- default delay time set to 0.1s")`

File: typeracer_bot.py [hunk 7]
```diff
@@ -1,78 +1,81 @@
 import time
+from contextlib import suppress
+
 from bs4 import BeautifulSoup
 from pynput.keyboard import Controller
 
-from selenium import webdriver
-from selenium.common.exceptions import TimeoutException
+from selenium.webdriver import Chrome, Firefox, Safari
+from selenium.webdriver.remote.webdriver import WebDriver
```
Comment (kwangzera) [inline typeracer_bot.py:8]: why not `from selenium import webdriver`

File: typeracer_bot.py (hunk 5 above)
Comment (Togohogo1) [inline typeracer_bot.py:58]: lol clarify

File: typeracer_bot.py [hunk 8]
```diff
@@ -1,78 +1,81 @@
...
+            return drivers[preferred]()
+    for name, func in drivers.items():
+        with suppress(WebDriverException):
+            return func()
+    raise LookupError("Driver not found: {preferred=!r}")
+
+def main(link="https://play.typeracer.com", delay=0.1, driver=None):
+    if not isinstance(driver, WebDriver):
+        driver = get_driver(preferred=driver)
+    keyboard = Controller()
```
Comment (Togohogo1) [inline typeracer_bot.py:63]: lol clarify

File: typeracer_bot.py [hunk 9]
```diff
@@ -1,78 +1,81 @@
...
+    if preferred is not None:
+        with suppress(WebDriverException):
+            return drivers[preferred]()
+    for name, func in drivers.items():
+        with suppress(WebDriverException):
+            return func()
+    raise LookupError("Driver not found: {preferred=!r}")
+
+def main(link="https://play.typeracer.com", delay=0.1, driver=None):
+    if not isinstance(driver, WebDriver):
```
Comment (kwangzera) [inline typeracer_bot.py:62]: can't u just do `if not driver`

File: typeracer_bot.py [hunk 10]
```diff
@@ -1,78 +1,81 @@
...
+# dict[name: lambda -> WebDriver]
+DRIVER_FUNCS = dict(
+    chrome=lambda: Chrome(executable_path="driver/chromedriver"),
+    firefox=lambda: Firefox(executable_path="driver/geckodriver"),
+    safari=lambda: Safari(executable_path="driver/safaridriver"),
+)
+
+def get_driver(preferred=None, drivers=DRIVER_FUNCS):
+    """Return first working driver in `drivers` dict."""
+    if preferred is not None:
```
Comment (kwangzera) [inline typeracer_bot.py]: why not just `if preferred`

File: typeracer_bot.py (hunk 3 above)
Comment (GeeTransit) [reply typeracer_bot.py]: It'll look weird. Plus, it might be useful to outside functions.

File: typeracer_bot.py (hunk 4 above)
Comment (GeeTransit) [reply typeracer_bot.py:1]: lmao i'll make another pull request to restore my way

File: typeracer_bot.py (hunk 7 above)
Comment (GeeTransit) [reply typeracer_bot.py:8]: Too much code. Compare `webdriver.Chrome` with `Chrome`. It gets worse with `webdriver.remove.webdriver.WebDriver` vs `WebDriver`. (webdriver gets typed 3 times...)

File: typeracer_bot.py (hunk 10 above)
Comment (GeeTransit) [reply typeracer_bot.py]: That works too, but I find the complementing `preferred=None` and `preferred is not None` nice to see (and to understand).

File: typeracer_bot.py (hunk 5 above)
Comment (GeeTransit) [reply typeracer_bot.py]: (O shoot I forgot the `f` before the string.) The `!r` applies `repr` over the result.

Example:
```py
>>> a = "derp"
>>> print(f"{a}")
derp
>>> print(f"{a!r}")
"derp"
>>> print(f"{a=}")
a=derp
>>> print(f"{a=!r}")
a="derp"
```

File: typeracer_bot.py (hunk 6 above)
Comment (GeeTransit) [reply typeracer_bot.py]: lol k i'll revert this.

File: typeracer_bot.py (hunk 9 above)
Comment (GeeTransit) [reply typeracer_bot.py:62]: `driver` can be a string (such as `"chrome"`).

File: typeracer_bot.py [hunk 11]
```diff
@@ -1,78 +1,82 @@
...
+        with suppress(WebDriverException):
+            return drivers[preferred]()
+    for name, func in drivers.items():
```
Comment (kwangzera) [inline typeracer_bot.py:56]: how about `for _, func in drivers.items():`

File: typeracer_bot.py (hunk 3 above)
Comment (kwangzera) [reply typeracer_bot.py]: how about defining it somewhere not in the middle of functions and also vscode gives me this warning `Dangerous default value DRIVER_FUNCS (builtins.dict) as argumentpylint(dangerous-default-value)`

File: typeracer_bot.py (hunk 11 above)
Comment (kwangzera) [reply typeracer_bot.py:56]: Better idea: `for func in drivers.values():`

File: typeracer_bot.py (hunk 7 above)
Comment (kwangzera) [reply typeracer_bot.py:8]: Won't be needing line 8 (this line) if I do `if not driver` on line 61. Also realized that `webdriver` and `WebDriver` are different things

File: typeracer_bot.py (hunk 3 above)
Comment (GeeTransit) [reply typeracer_bot.py]: lol The warning is because it's a dictionary. I'll move it to just below the imports

File: typeracer_bot.py (hunk 11 above)
Comment (GeeTransit) [reply typeracer_bot.py:56]: Ye that was left over from when I was gonna raise an exception using the name of the driver.

File: typeracer_bot.py (hunk 9 above)
Comment (GeeTransit) [reply typeracer_bot.py:62]: `driver` can have 3 possible types: `None`, `str`, or `WebDriver`.

When it's `None`, it'll go through the drivers like normal (checking if it works). `str` also sends the preferred driver to `get_driver`, meaning it'll attempt that driver first, then continue just like `None`. When a `WebDriver` is passed, it uses it directly, skipping the `get_driver`.

The code needs to check if the driver is a WebDriver (to skip or not to skip), so using `not isinstance(driver, WebDriver)` would work. If the check used `not driver`, a string would be used as the driver (which doesn't work).

**Pull request merged**